Restructure crate to contain lib and bin

anxiousmodernman opened the pull request:
The binary name will be the same. The binary can be built like

```
cargo build --bin fitbit-grabber
```

The new library name is "fitbit", and src/bin/main.rs imports it. We could
configure to have more binaries, but only one top-level library.

Error handling:
Convert errors to failure crate.

On the library side, wrap errors from our dependencies with an enum, and derive
Fail for it. This approach to lets us introduce strong types with custom
printing fairly quickly. It has the disadvantage of being somewhat rigid, and
is one of a few approaches described here:
https://boats.gitlab.io/failure/custom-fail.html

On the binary side, as a user of the library, failure allows us to make liberal
use of the `?` operator, even in main, since main can do `Result` now.

Config changes, allow a config file:
Create application config module

config::Config can be deserialized from a config file at

~/.config/fitbit-grabber/conf.toml

or set with --config at the command line.

The config struct is all Options that must be unwrapped, so we could have
separate command funcs that care about particular values on a Config.

anxiousmodernman committed:
Restructure crate to contain lib and bin

The binary name will be the same. The binary can be built like

```
cargo build --bin fitbit-grabber
```

The new library name is "fitbit", and src/bin/main.rs imports it. We could
configure to have more binaries, but only one top-level library.

Error handling:
Convert errors to failure crate.

On the library side, wrap errors from our dependencies with an enum, and derive
Fail for it. This approach to lets us introduce strong types with custom
printing fairly quickly. It has the disadvantage of being somewhat rigid, and
is one of a few approaches described here:
https://boats.gitlab.io/failure/custom-fail.html

On the binary side, as a user of the library, failure allows us to make liberal
use of the `?` operator, even in main, since main can do `Result` now.

Config changes, allow a config file:
Create application config module

config::Config can be deserialized from a config file at

~/.config/fitbit-grabber/conf.toml

or set with --config at the command line.

The config struct is all Options that must be unwrapped, so we could have
separate command funcs that care about particular values on a Config.

anxiousmodernman commented on src/lib.rs:
```diff
@@ -0,0 +1,341 @@
+        })
+    }
+
+    pub fn user(&self) -> Result<String, Error> {
```
The various data-fetching methods can be moved to traits.

anxiousmodernman commented on src/bin/main.rs:
```diff
@@ -0,0 +1,133 @@
+        let date = NaiveDate::parse_from_str(&raw_date, "%Y-%m-%d")
+            .map_err(|e| format_err!("could not parse date {}", e))?;
+        let step_data = f.step(date)?;
+        println!("{}", step_data);
```
This method, `step`, still currently returns a string. If we model it to return library structs, we will need to re-serialize json here to preserve the same behavior as before. Serde might print slightly different after reserialization.

anxiousmodernman committed:
Rename crate: fitbit

This name is available on crates.io, but in any case we can use it since
no client lib exists.

bradfordboyle commented on Cargo.toml:
```diff
@@ -1,16 +1,26 @@
 serde_derive = "1.0"
 serde_json = "1.0"
 tiny_http = "0.6"
+toml = "*"
```
I think this dependency is for the CLI utility and not the library. Do you think we should separate the dependencies into target specific lists? We can create an issue and do that after this gets merged.

Should we have a version specifier?

anxiousmodernman replied:
It should have a version, for sure.

I actually backed out a lot more cli stuff, but left this in since it didn't seem disruptive, and perhaps useful. I made a new repo for a more daemony thing, and specified _this_ repo as a local path dependency: https://github.com/anxiousmodernman/fitbit-syncd/blob/master/Cargo.toml#L26

In the long run, separate crates for bins and libs is probably the way to go. But we can keep things co-mingled here for a while.

bradfordboyle commented on src/bin/config.rs:
```diff
@@ -0,0 +1,76 @@
+}
+
+impl Config {
+    /// Load a config from the environment. A config object will be constructed
```
👍

bradfordboyle commented on src/bin/config.rs:
```diff
@@ -0,0 +1,76 @@
+            }
+            // default path
+            None => {
+                let base_dir = env::var("HOME")?;
```
What about using [`env::home_dir()`](https://doc.rust-lang.org/std/env/fn.home_dir.html)?

bradfordboyle commented on src/bin/config.rs:
```diff
@@ -0,0 +1,76 @@
+            // custom path to config file, passed as a flag
+            Some(path) => {
+                //if let Some(found) = try_file(&PathBuf::from(Path::new(pathbuf))) {
+                if let Ok(found) = Config::from_toml_file(path) {
```
The code for reading the config in from a file is the same so I'm wondering if it makes sense to restructure this a bit

```rust
let config_path = match path {
    Some(path) => path,
    None => { ... }
}

let config = Config::from_toml_file(config_path)?;
```

It also seems like this is silently ignoring the case where there is a config file but it is not valid TOML.

bradfordboyle commented on src/bin/main.rs:
```diff
@@ -0,0 +1,133 @@
+fn main() -> Result<(), Error> {
+    let default_dir = Path::new(&env::var("HOME")?).join(".config/fitbit-grabber");
+    let default_config = default_dir.clone().join("conf.toml");
+    let date_arg = Arg::with_name("date")
```
👍

bradfordboyle commented on src/lib.rs:
```diff
@@ -0,0 +1,341 @@
+}
+
+pub trait User {
+    fn get_profile(&self) -> Result<UserProfileResult, Error>;
```
I really like splitting the different sections into different traits. What about splitting these into separate modules?
```rust
// src/lib.rs
pub use self::user::User;

mod user;

// src/user.rs
pub trait User {...}

impl User for FitbitClient {...}

// src/main.rs
use fitbit_grabber::User;

```

How do you feel about naming this method `get_user_profile()`? I think it would be good to try and have the method names match the URL endpoints. At one point, I had thought about trying to use [service objects](https://medium.com/@marcus.olsson/writing-a-go-client-for-your-restful-api-c193a2f4998c) so that the API would look like

```rust
client.user.profile()
```

but I needed the client to own the user and the user to have a reference back to the client and my rust-fu was weak.

anxiousmodernman replied:
I agree on that naming style moving forward. That feels like how their docs read, to me.

I also agree with putting each trait into a same-named module. From what I can tell it should all map pretty nicely.

bradfordboyle commented on src/lib.rs:
```diff
@@ -0,0 +1,341 @@
+impl User for FitbitClient {
+    fn get_profile(&self) -> Result<UserProfileResult, Error> {
+        let url = self.base.join("user/-/profile.json")?;
+        Ok(self
```
I think this code will be repeated in a lot of places. Leaving it "stringly"-typed we could extract a helper method like

```rust
fn get_profile(&self) -> Result<???> {
    let url = self.base.join(...)
    self.client.do_get(url)
}

impl FitbitClient {
    fn do_get(&self, path: &str) -> Result<String> {
        let url = self.base.join(&path)?;
        self.client
            .request(Method::Get, url)
            .send()
            .and_then(|mut r| r.text())
    }
}
```

With the strongly-typed API, I'd have to think about how to do this method extraction. Do you have any ideas?

anxiousmodernman replied:
yeah, a helper would be good here; (hand-waving) something like `do_get<T>(url: &str) -> Result<T, Error>`.  There must be a way to make this generic plumbing for what eventually is serde `Deserialize`.

bradfordboyle commented on src/lib.rs:
```diff
@@ -0,0 +1,341 @@
+    base: url::Url,
+}
+
+pub enum DateQuery {
```
👍

bradfordboyle commented on src/lib.rs:
```diff
@@ -0,0 +1,341 @@
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct UserProfileResult {
```
The stuct above makes sense to me; I'm not sure I'm seeing the reason for wrapping into another struct.

anxiousmodernman replied:
This is a mapping of their nested response
```
{
  "user": { tons of fileds }
}
```
Serde wants each object to be mapped precisely. We could avoid this by deserializing something like `HashMap<String, UserProfile>`, and looking up the sub-object.

Either way, they have a lot of objects with this shape in Fitbit. I'd be down to hide the wrapper types from the public api, come to think of it, although we might need them for deserialization. On the other hand, future library users might expect us to map the Json response faithfully - even if it's kind of cludgy.

bradfordboyle commented on src/lib.rs:
```diff
@@ -0,0 +1,341 @@
+
+#[cfg(test)]
+mod tests {
+    // NOTE: where's this from? Tests don't compile.
```
Yeah, this is dead code and I need to remove it.

bradfordboyle reviewed (approved):
This is awesome! I've made a few comments (mostly discussion questions) but see no real issue with merging this.

anxiousmodernman commented on Cargo.toml:
```diff
@@ -1,16 +1,26 @@
 authors = ["Bradford D. Boyle <bradford.d.boyle@gmail.com>"]
+autobins = false
 license = "Apache-2.0"
+name = "fitbit"
```
PS before you merge: I added this name change. It looks nicer when importing, and you don't need to claim the spot on crates.io yet

anxiousmodernman committed:
pin toml version

anxiousmodernman committed:
simplify code

anxiousmodernman committed:
Named home_dir function

This returns an Option, so we map it.

anxiousmodernman committed:
Simplify code more

and_then is convenient for debugging the http headers, but adds a lot of code.
We might want to keep a stringly typed debug method around for dev.

anxiousmodernman committed:
user_profile->get_user_profile
